File: skills/fastmail/scripts/jmap-client.ts
interface JMAPSession {
  apiUrl: string;
  downloadUrl: string;
  uploadUrl: string;
  accounts: Record<string, {
    name: string;
    isPersonal: boolean;
    accountCapabilities: Record<string, unknown>;
  }>;
  primaryAccounts: Record<string, string>;
}

interface JMAPRequest {
  using: string[];
  methodCalls: [string, Record<string, unknown>, string][];
}

interface JMAPResponse {
  methodResponses: [string, Record<string, unknown>, string][];
  sessionState: string;
}

export class JMAPClient {
  private session: JMAPSession | null = null;
  private token: string;
  
  constructor(token: string) {
    this.token = token;
  }
  
  async getSession(): Promise<JMAPSession> {
    if (this.session) return this.session;
    
    const response = await fetch('https://api.fastmail.com/jmap/session', {
      headers: {
        'Authorization': `Bearer ${this.token}`,
      },
    });
    
    if (!response.ok) {
      throw new Error(`Failed to get session: ${response.statusText}`);
    }
    
    this.session = await response.json();
    return this.session!;
  }
  
  async call(methodCalls: [string, Record<string, unknown>, string][]): Promise<JMAPResponse> {
    const session = await this.getSession();
    
    const request: JMAPRequest = {
      using: [
        'urn:ietf:params:jmap:core',
        'urn:ietf:params:jmap:mail',
        'urn:ietf:params:jmap:submission',
      ],
      methodCalls,
    };
    
    const response = await fetch(session.apiUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.token}`,
      },
      body: JSON.stringify(request),
    });
    
    if (!response.ok) {
      throw new Error(`JMAP call failed: ${response.statusText}`);
    }
    
    return response.json();
  }
  
  async getAccountId(): Promise<string> {
    const session = await this.getSession();
    return session.primaryAccounts['urn:ietf:params:jmap:mail'];
  }
}

File: skills/fastmail/scripts/jmap_client.py
"""JMAP client for Fastmail email operations."""

import requests


class JMAPClient:
    SESSION_URL = "https://api.fastmail.com/jmap/session"

    def __init__(self, token: str):
        self._token = token
        self._session: dict | None = None

    def _headers(self) -> dict:
        return {"Authorization": f"Bearer {self._token}"}

    def get_session(self) -> dict:
        if self._session:
            return self._session

        resp = requests.get(self.SESSION_URL, headers=self._headers())
        resp.raise_for_status()
        self._session = resp.json()
        return self._session

    def call(self, method_calls: list[tuple[str, dict, str]]) -> dict:
        session = self.get_session()

        request_body = {
            "using": [
                "urn:ietf:params:jmap:core",
                "urn:ietf:params:jmap:mail",
                "urn:ietf:params:jmap:submission",
            ],
            "methodCalls": method_calls,
        }

        resp = requests.post(
            session["apiUrl"],
            json=request_body,
            headers={**self._headers(), "Content-Type": "application/json"},
        )
        resp.raise_for_status()
        return resp.json()

    def get_account_id(self) -> str:
        session = self.get_session()
        return session["primaryAccounts"]["urn:ietf:params:jmap:mail"]
